Adding multi-server support for pbs_deljoblist

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
Enhancing pbs_deljoblist to add support for multi-server, plus some refactoring.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
qdel first passes IFL the list of jobs until the mail limit, and then it passes IFL the entire list of jobs with "nomail" direction, for each server. IFL takes care of grouping them by each multi-server instance and batching them into deljoblist requests. qdel program still handles jobs submitted to/moved to remote clusters.

Why all the refactoring?:
- Keeping multi-server related code (grouping jobs for each instance, finding fd of each mutli-server, broadcasting etc.) in IFL only is cleaner than it being split between client program and IFL
- allows clients to use the IFL layer directly instead of the client program, which is consistent with the overall design for how IFL is supposed to behave.
- It also makes it easier for one to add customized IFL functions (via IFL function pointers) as the custom IFL won't have to support the client program wanting to do multi-server specific stuff.

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
**Deleting jobs submitted to multiple servers**
```
[ravi@pbsc8 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
87065.pbsc8       STDIN            ravi                     0 Q workq           
87066.pbsc8       STDIN            ravi                     0 Q workq           
87067.pbsc8       STDIN            ravi                     0 Q workq           
4.pbsc8-2         STDIN            ravi                     0 Q workq           
5.pbsc8-2         STDIN            ravi                     0 Q workq           
6.pbsc8-2         STDIN            ravi                     0 Q workq           
7.pbsc8-2         STDIN            ravi                     0 Q workq           
8.pbsc8-2         STDIN            ravi                     0 Q workq           
9.pbsc8-2         STDIN            ravi                     0 Q workq           
10.pbsc8-2        STDIN            ravi                     0 Q workq           
11.pbsc8-2        STDIN            ravi                     0 Q workq           
[ravi@pbsc8 ~]$ qdel `qselect`
[ravi@pbsc8 ~]$ qstat
[ravi@pbsc8 ~]$ 
```

**Deleting jobs running on multiple servers, with move + runs**:
```
[ravi@pbsc8 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
32.pbsc8-2        STDIN            ravi              00:00:00 R workq           
33.pbsc8-2        STDIN            ravi              00:00:00 R workq           
34.pbsc8-2        STDIN            ravi              00:00:00 R workq           
35.pbsc8-2        STDIN            ravi              00:00:00 R workq           
45.pbsc8          STDIN            ravi              00:00:00 R workq           
46.pbsc8          STDIN            ravi              00:00:00 R workq           
47.pbsc8          STDIN            ravi              00:00:00 R workq           
36.pbsc8-2        STDIN            ravi              00:00:00 R workq           
37.pbsc8-2        STDIN            ravi                     0 Q workq           
38.pbsc8-2        STDIN            ravi                     0 Q workq           
39.pbsc8-2        STDIN            ravi                     0 Q workq           
[ravi@pbsc8 ~]$ 
[ravi@pbsc8 ~]$ 
[ravi@pbsc8 ~]$ pbsnodes -av
pbsc8
     Mom = pbsc8
     ntype = PBS
     state = job-busy
     pcpus = 4
     jobs = 32.pbsc8-2/0, 33.pbsc8-2/1, 34.pbsc8-2/2, 35.pbsc8-2/3
     resources_available.arch = linux
     resources_available.host = pbsc8
     resources_available.mem = 2035188kb
     resources_available.ncpus = 4
     resources_available.vnode = pbsc8
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 4
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     license = l
     last_state_change_time = Tue Dec 15 07:04:32 2020
     last_used_time = Tue Dec 15 07:04:23 2020

pbsc8-2
     Mom = pbsc8
     ntype = PBS
     state = job-busy
     pcpus = 4
     jobs = 45.pbsc8/0, 46.pbsc8/1, 47.pbsc8/2, 36.pbsc8-2/3
     resources_available.arch = linux
     resources_available.host = pbsc8
     resources_available.mem = 2035188kb
     resources_available.ncpus = 4
     resources_available.vnode = pbsc8-2
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 4
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     license = l
     last_state_change_time = Tue Dec 15 07:04:34 2020
     last_used_time = Tue Dec 15 07:04:23 2020

[ravi@pbsc8 ~]$ 
[ravi@pbsc8 ~]$ 
[ravi@pbsc8 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
32.pbsc8-2        STDIN            ravi              00:00:00 R workq           
33.pbsc8-2        STDIN            ravi              00:00:00 R workq           
34.pbsc8-2        STDIN            ravi              00:00:00 R workq           
35.pbsc8-2        STDIN            ravi              00:00:00 R workq           
45.pbsc8          STDIN            ravi              00:00:00 R workq           
46.pbsc8          STDIN            ravi              00:00:00 R workq           
47.pbsc8          STDIN            ravi              00:00:00 R workq           
36.pbsc8-2        STDIN            ravi              00:00:00 R workq           
37.pbsc8-2        STDIN            ravi                     0 Q workq           
38.pbsc8-2        STDIN            ravi                     0 Q workq           
39.pbsc8-2        STDIN            ravi                     0 Q workq           
[ravi@pbsc8 ~]$ qdel `qselect`
[ravi@pbsc8 ~]$ 
[ravi@pbsc8 ~]$ qstat
[ravi@pbsc8 ~]$ 
```

**Deleting 50k jobs:**
```
[ravi@pbsc8 ~]$ qstat | wc -l
50002
[ravi@pbsc8 ~]$ qdel `qselect`
[ravi@pbsc8 ~]$ qstat
```

**Performance number for qdel of 50k jobs:**
```
[ragrawal@* debug]$ qstat | wc -l
50002
[ragrawal@blrentperf04 debug]$ time qdel `qselect`

real	0m12.038s
user	0m0.146s
sys	0m0.052s
```

**Qdel of jobs submitted to multiple clusters:**
```
[ragrawal@m2 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
2.m2              STDIN            ragrawal                 0 Q workq           
3.m2              STDIN            ragrawal                 0 Q workq           
[ragrawal@m2 ~]$ qstat @m1
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
2.m1              STDIN            ragrawal                 0 Q workq           
3.m1              STDIN            ragrawal                 0 Q workq           
4.m1              STDIN            ragrawal                 0 Q workq           
[ragrawal@m2 ~]$ qdel 2.m1 2.m2
[ragrawal@m2 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
3.m2              STDIN            ragrawal                 0 Q workq           
[ragrawal@m2 ~]$ qstat @m1
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
3.m1              STDIN            ragrawal                 0 Q workq           
4.m1              STDIN            ragrawal                 0 Q workq 
```
**Qdel of jobs moved to remote cluster:**
```
[ragrawal@m2 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
3.m2              STDIN            ragrawal                 0 Q workq           
[ragrawal@m2 ~]$ qstat @m1
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
3.m1              STDIN            ragrawal                 0 Q workq           
4.m1              STDIN            ragrawal                 0 Q workq           
[ragrawal@m2 ~]$ qmove workq@m1 3.m2
[ragrawal@m2 ~]$ qstat @m2
[ragrawal@m2 ~]$ qstat @m1
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
3.m1              STDIN            ragrawal                 0 Q workq           
4.m1              STDIN            ragrawal                 0 Q workq           
3.m2              STDIN            ragrawal                 0 Q workq           
[ragrawal@m2 ~]$ qdel 3.m1 3.m2
[ragrawal@m2 ~]$ qstat
[ragrawal@m2 ~]$ qstat @m1
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
4.m1              STDIN            ragrawal                 0 Q workq  
```

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

agrawalravi90 commented:
@nithinj and @bremanandjk  requesting review, thanks.

nithinj commented:
@agrawalravi90 ,

You have explained that you have moved the grouping from the client to the IFL layer. But it is still achieved in qdel as per code. Could you update the description if it is not up to date?

Also, please explain a bit more about refactoring and why are you doing it. It looks like most of the changes are related to refactoring and not making pbs_deljoblist to work with multi-server.

agrawalravi90 commented:
> @agrawalravi90 ,
You have explained that you have moved the grouping from the client to the IFL layer. But it is still achieved in qdel as per code. Could you update the description if it is not up to date?
Also, please explain a bit more about refactoring and why are you doing it. It looks like most of the changes are related to refactoring and not making pbs_deljoblist to work with multi-server.

Thanks for pointing that out, I've updated the description to add more details, please let me know if you have any questions.

nithinj commented:
All the other IFL changes related to multi-server were done just within IFL so that clients do not have to know about the inernals. It seems deljoblist also can be made to work in the same way:
1. Broadcast the request in case of multi-svr
2. Collate the results before sending back to the client

But this change is a bit extensive and does touch the client code a lot. So making it clear on why a lot of refactoring was required would help.

It appears that the way this PR has handled IFL is different from how  job-related IFL is getting handled in https://github.com/openpbs/openpbs/pull/2086 . It needs to be unified to avoid duplication of code and make maintenance easy.

agrawalravi90 commented:
> It seems deljoblist also can be made to work in the same way:
Broadcast the request in case of multi-svr
Collate the results before sending back to the client

Broadcast the list of all jobs to delete to all instances of multi-server? That sounds really bad for performance, it will unnecessarily hold up the all servers.

> It appears that the way this PR has handled IFL is different from how job-related IFL is getting handled in #2086 . It needs to be unified to avoid duplication of code and make maintenance easy.

The approach is the same right? you are also first trying to contact the parent server of the job and then broadcasting it to all servers. The main difference is that you are being a bit more rigorous in finding the owner server and I'm broadcasting the request in parallel while you are doing it serially. With jobid namespace redesign, we'll be able to find out owner much more easily, so we can refactor all IFLs to use that common code when we add it. We could also add common code for broadcasting, but I'd prefer my parallel approach for it. Anyways, all this is just code cleanup which can be done later.

nithinj commented:
@agrawalravi90 ,
I was referring to functions you have added like server_name_to_fd() and get_server_fd_from_jid() which are also present in the other PR in some other forms. I am holding from reviewing them as it will have to be unified.
Also, I assume the test you have shown should be working today with multi-server as jobs are restricted to server instances and qdel already groups them by server. It needs to be tested after move and run checks in and when jobs moves around.

agrawalravi90 commented:
> I was referring to functions you have added like server_name_to_fd() and get_server_fd_from_jid() which are also present in the other PR in some other forms. I am holding from reviewing them as it will have to be unified.

Not really, the other PR adds a function which returns a server index, not fd, so I don't think get_server_fd_from_jid() is a duplicate. I also don't think that 'server_name_to_fd' is duplicated. The way that they get to their answer is duplicated, yes, so we can refactor that and put it into a common function. There's no point creating that common function now because it will be obsolete once we redesign jobids. So, why not wait until then to refactor code? 

> Also, I assume the test you have shown should be working today with multi-server as jobs are restricted to server instances and qdel already groups them by server. It needs to be tested after move and run checks in and when jobs moves around.

No, it doesn't. Let's say you have 2 jobs, 1.server1, 1.server2, existing qdel calls groups them separately, true, but when it connects with 'server1' and 'server2', it is connecting to the same host, on the same port (because it doesn't know any better). So, it sends both the jobs to the same server:
```
[ravi@pbsc8 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
3.pbsc8           STDIN            ravi                     0 Q workq           
4.pbsc8           STDIN            ravi                     0 Q workq           
6.pbsc8-2         STDIN            ravi                     0 Q workq           
7.pbsc8-2         STDIN            ravi                     0 Q workq           
[ravi@pbsc8 ~]$ 
[ravi@pbsc8 ~]$ qdel 3.pbsc8 6.pbsc8-2
[ravi@pbsc8 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
4.pbsc8           STDIN            ravi                     0 Q workq           
6.pbsc8-2         STDIN            ravi                     0 Q workq           
7.pbsc8-2         STDIN            ravi                     0 Q workq  
```

arungrover commented on src/cmds/qdel.c:
```diff
@@ -58,106 +58,212 @@
+		fd = pbs_connect(iter_remote->svrname);
+		if (fd > 0) {
+			p_delstatus = pbs_deljoblist(connect, iter_remote->jobids, iter_remote->total_jobs, warg);
+			any_failed = process_deljobstat(clusterid, p_delstatus, NULL);
```
Here as well, do we mean to overwrite the any_failed field?

arungrover commented on src/cmds/qdel.c:
```diff
@@ -58,106 +58,212 @@
+ * @param[in]	numofjobs - count of jobs to delete
  *
+ * @return int
+ * @retval 0 for Success
```
Since you are returning any_failed value, it could potentially be an error code, no?

agrawalravi90 commented on src/cmds/qdel.c:
```diff
@@ -58,106 +58,212 @@
+			errtxt = pbse_to_txt(p_delstatus->code);
+			if ((errtxt != NULL) && (p_delstatus->code != PBSE_HISTJOBDELETED)) {
+				fprintf(stderr, "%s: %s %s\n", "qdel", errtxt, p_delstatus->name);
+				any_failed = p_delstatus->code;
```
Yes, that's what the existing code was doing (i.e - return the code of the last job that failed)

agrawalravi90 commented on src/cmds/qdel.c:
```diff
@@ -58,106 +58,212 @@
+			exit(pbs_errno);
+		}
+
+		for (iter_ss = ss; iter_ss != NULL; iter_ss = iter_ss->next) {
```
> I assume this loop is to loop on multiple servers.
If that is the case then shouldn't we be setting dfltmail for each corresponding server? Can't two different servers have different values of default qdel args?

No I think this loop was a mistake, we will never have more than one object returned by statserver. Also, we use 'ss' instead of 'iter_ss' inside the loop so it was only looking at the first member of the list.

I call this function from "delete_jobs_for_cluster", which is called for every cluster (or every server, pre multi-server), so it should consider each cluster's unique dflt_qdel_args.

agrawalravi90 commented:
@arungrover thanks for the comments, I've addressed them, please provide further feedback

nithinj reviewed (changes requested):
Post the logs with jobs in the running state while deleting with multi-sevrer.

arungrover commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,471 @@
+ *
+ * @param[in]	jobid - id of the job
+ * @param[in]	code - error code
+ * @param[in]	list - list to add new object to
```
This does not match with the argument

arungrover commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,471 @@
+add_failed_job(char *jobid, int code, struct batch_deljob_status **ret)
+{
+	struct batch_deljob_status *djstat = NULL;
+
```
Please add a check for ret to be non-null

agrawalravi90 replied:
for static functions I'm usually not as strict with such checks because it's kind of like an 'internal' function to that file, it exists for the sake of creating multiple smaller functions instead of one big function. But if you think it's important then I'll add it, please let me know.

arungrover commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,471 @@
+		if (!skip_list[i]) {
+			struct batch_deljob_status *ret = NULL;
+
+			ret = recv_deljob(conns[i].sd);
```
How does this work? You are sending a list of job ids to all other servers, what do we expect in recv_deljob? does each server reply per job or as a whole for the list of jobs?

agrawalravi90 replied:
Each server replies only once, for the whole list of jobs. So, for 5 servers, we'll send 5 requests out and get 5 responses back

agrawalravi90 commented on src/include/cmds.h:
```diff
@@ -56,6 +56,17 @@
 #include "libsec.h"
 
+/* Needed for qdel and pbs_deljoblist */
+typedef struct svr_jobid_list svr_jobid_list_t;
```
why?

agrawalravi90 commented on src/lib/Libifl/pbsD_deljob.c:
```diff
@@ -68,15 +70,17 @@
+	list[0] = jobid;
+	list[1] = NULL;
+
+	res = __pbs_deljoblist(c, list, 1, extend);
```
i think it's better to have only one codepath for job deletion. That way, any enhancements done to deljoblist can be used by deljob. If there's a different code path for job deletion then we should probably remove it.

nithinj replied:
I would suggest removing any dead code created as part of making this change. You can defer it but both should happen together.
Multi-svr specific changes are already handled for pbs_deljob as it is calling PBSD_manager internally.

agrawalravi90 replied:
deljoblist added another path to deleting jobs, so we should remove the path that exists via manager. This PR removes the client to server path. I think the server itself uses the manager path to talk to mom, removing that seems out of scope for this PR, we can take it up separately.

nithinj replied:
There should be code associated with what you have removed from the client side in req_deletejob(). That becomes dead code as part of this change. If you are removing the client path, remove the same from the server as well.

If you are seeing it as out of scope, refrain from updating the client now and take it up together.

agrawalravi90 replied:
the server still uses the old request type for various things, for example, it uses it to send delete job request to remote servers in case the job was moved, it uses it for preemption by deletion, for deleting jobs when a reservation ends, and to tell mom to delete a job, i might be missing other cases. So, certainly not dead code just because I changed the client to not use it. We should change the server to not use it, yes, but it's an internal refactoring which needs wide spanning changes in multiple daemons, completely out of scope for this PR.

agrawalravi90 commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,471 @@
+	svr_jobid_list_t *svr_jobid_list_hd = NULL;
+	int num_servers = get_num_servers();
+
+	if (num_servers > 1) {
```
> Check the output of get_conn_svr_instances() to arrive at this decision. get_num_servers() is also not relevant for remote clusters.

Why is that? get_num_servers() should return 1 for a single server (remote or local), right? We should make these functions as generic as possible for both multi and single server setups.

nithinj replied:
* PBSD_deljoblist() can be used to invoke with an instance fd if the client knows where the job belongs to. In that case, get_num_servers() will misguide you and say its a multi-svr cluster although you could use the single server code path.
* get_num_servers() can be made generic for both remote and local if you pass some argument to it. It should be able to distinguish what you are calling it for. For that matter get_conn_svr_instances() is already being invoked on an fd and it can return the right number of fd based on what is passed to it.
I would suggest the following to make the same decision:
```
def is_msvr(fd) {
  conns = get_conn_svr_instances(c);
  if (!conns || conns[0]->sd == fd)
          return FALSE
  return TRUE
}
```

agrawalravi90 replied:
>  In that case, get_num_servers() will misguide you and say its a multi-svr cluster although you could use the single server code path.

Why though? get_num_servers() should return 1 in case of a single server, so a simple check for > 1 should be enough to know whether we are in single or multi-server setup, right? I think I'm missing something.

agrawalravi90 replied:
oh wait, i think i get what you mean, even in a multi-server setup, one might call this function with a single server's fd .. let me take care of it

agrawalravi90 commented:
@nithinj  I've addressed your comments, it involved some major changes to broadcast, so i pushed them before addressing @arungrover 's comments, please take a look. I'll address @arungrover 's comments soon, thanks

agrawalravi90 commented:
Also, is it ok to rebase this with master? it makes it a pain to test with Nithin's move + run changes.

agrawalravi90 commented:
@arungrover  and @nithinj  thanks guys, I've rebased with master, please resume the review process.

nithinj commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,554 @@
+	} else {	/* Single server */
+		svr_conn_t **conns = get_conn_svr_instances(c);
+
+		if (conns == NULL) {
```
This could be the case even if the caller passes an instance fd

nithinj commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,554 @@
+	struct batch_deljob_status *prev = NULL;
+	struct batch_deljob_status *next = NULL;
+
+	if (conns == NULL) {
```
get_conn_svr_instances will return NULL for instance fd.

agrawalravi90 replied:
We will not broadcast for an instance fd case. So, once inside this function, conns should not be NULL.

arungrover commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,554 @@
+			if (prev != NULL)
+				prev->next = next;
+			else
+				svr_retlist[0] = next;
```
Nithin has a good point. Do we really have to report users about jobs that are not present in any of the servers? Can we just not iterate through the lists and report jobs that servers fail to delete because of other reasons? Those are the actual jobs a user would care about, isn't it?

agrawalravi90 replied:
I agree, we should care more about legit failures to delete the job. But here, we are removing the jobids which were successfully deleted on other servers, so it's perfectly safe to ignore their error message. qdel is returned the complete list of all jids with error codes if they couldn't be deleted on any servers.

agrawalravi90 replied:
on second thought, to return the accurate error code, I'll need to add a check inside the loop above this, that's where we go through the lists of all servers, so thanks for pointing this out.

arungrover replied:
Here is what I understand from what I read, please correct me if I'm wrong - 
If you have 'm' servers and they reply with 'n' number of jobs that couldn't be deleted, you are looping through all servers and the jobs they reported, so the complexity is close to m*n^2.
Can't you just loop through all the lists one after another and just gather jobs that have code other than unknown job id? that will result in lowering down complexity to m*n, isn't it? I think it is safe to do as long as you don't want to return back jobs with unknown job error code.

nithinj replied:
I agree with the approach. Just noting that we also might need to report back unknown job id, if that is what you heard back from all the servers.

agrawalravi90 replied:
that's the thing, we do need to report back unknowns as well if no server found the job. The only way to know if a job was found on others servers or not is by scanning each server's list, which adds that extra n and makes it mn^2. @nithinj did you have an algorithm in mind which reduces the time complexity?

In practice, the broadcast list should be small, we anticipate a small fraction of jobs to be moved between servers, so I don't think we have to worry about this, but if there's an obviously more performant algorithm, please do let me know.

nithinj replied:
You could use avl indexing to reduce complexity. But that seems unnecessary if the number of jobs is going to be a small percentage. Also, deljoblist is batching jobs before sending them to the server. 
One optimization you can do in your current approach is to break the inner loop of iterating over nsvrs when you encounter an rc != UNKWN_ID or when you don't find the job in a servers list.

agrawalravi90 replied:
"You could use avl indexing to reduce complexity" I agree with you that it's not worth it, but I did think about it and didn't really see how it would help, so out of curiosity, can you please tell me what you had in mind?

"One optimization you can do in your current approach is to break the inner loop of iterating over nsvrs when you encounter an rc != UNKWN_ID or when you don't find the job in a servers list."
ok ya, I should definitely do that, thanks for the suggestion

nithinj replied:
You need to walk over all the jobs returned by servers anyway. That is mn.
When we encounter a new jobid, add it to the index.
If the index contains the ID, update it with rc and the number of times it had been repeated.
Before returning the final list remove every element which has not repeated m times and having rc as unknwn_ID.
Since we are accessing from index for every m*n, the complexity will be mnlogn

arungrover replied:
do we have to report unknown job ids? qdel in its current form does not report that.

agrawalravi90 replied:
oh, that's interesting. If we don't need to then we can simplify this. Thanks!

nithinj replied:
Your process_deljobstat function has this check. If deljoblist returns PBSE_UNKJOBID, which means the job got moved to a different cluster. It sends a locatejob to identify where the job has been moved to and sends a delete request to that cluster.

bremanandjk replied:
Yeah, The current qdel client would need to know the return code PBSE_UNKJOBID for the moved job, so that qdel would locate it and send the delete request.

agrawalravi90 replied:
oh right, ill need to revert this change then, thanks for catching that guys

arungrover commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,554 @@
+ *
+ * @param[in] c - connection handler
+ * @param[in] jobid - job identifier
+ * @param[in] extend - string to encode req
```
you are missing numjids

agrawalravi90 commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,554 @@
+	}
+
+	/* Broadcast the jids to all servers */
+	for (i = 0; i < num_svrs; i++) {
```
Well I need to call get_num_servers() anyways in that function, and some of the loops need it instead of conns[i] != NULL, so I'd prefer being consistent and using i < num_svrs everywhere.

nithinj replied:
I do not see which loop need to use i < num_svrs in the changelist.
But if you are sticking with this, add a check against conns[i] to make sure that its not NULL.

agrawalravi90 commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,554 @@
- * @retval	0	success
- * @retval	!0	error
+ * @param[in] job_id - Job id
+ * @param[in] server - server name
```
I'm sorry for doing such a bad job at keeping the function descriptors consistent

arungrover replied:
hey no worries, we all do it :)

agrawalravi90 commented:
@arungrover  and @nithinj  ive addressed your comments, please provide further feedback, thanks

nithinj commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,554 @@
-		if ((rc = is_compose_cmd(c, IS_CMD, msgid)) != DIS_SUCCESS)
-			return rc;
-	}
+	if (fd_virtual) {	/* Virtual fd means we can talk to more than 1 server instance */
```
Not necessarily according to your definition of is_fd_virtual(). There can be just one instance or n instances based on the cluster configuration.
If it is just one, you might still want to take the optimal path of a single server.

agrawalravi90 replied:
ok, ive renamed the function to multi_svr_op() and am checking more than just get_conn_svr_instances() == NULL, let me know what you think

nithinj replied:
Although what you are going to perform is an operation, the check itself is not. Would it be better to call it multi_svr(fd) without the _op part?

agrawalravi90 replied:
We could be in multi-server setup but still the function will return False if fd is instance fd. So, I think the _op makes it a bit more clear that we are talking about single server operation vs multi-server operation, regardless of the setup. So, I'd prefer keeping it

nithinj commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,554 @@
+					next = iter_list->next;
+					if (iter_list->code == PBSE_UNKJOBID && fd_virtual) {
+						/* Add job to the broadcast list */
+						if (nbcastids >= (bcastjidssize - 1)) {
```
Can add jobs to broadcast list be a separate function. You can hide all the buffer reallocation etc inside the function. It will also make PBSD_deljoblist() more readable. Its 100+ lines now.

nithinj commented on src/lib/Libifl/pbsD_deljoblist.c:
```diff
@@ -46,160 +46,554 @@
+		int error_code = PBSE_UNKJOBID;
+
+		/* Make sure that this jid is in ALL of the other lists as well */
+		for (i = 1; i < num_svrs; i++) {
```
This is a bad assumption to make. pbs_deljoblist can be used to delete jobs from a remote cluster. Right now pbs_connect is not enhanced to talk to a remote msvr-cluster. But the moment we do that we have to revisit all these codes. Because num_svrs is local and not remote.
Instead if you had relied on the size of the connection array, all these will still be valid. But its up to you.

agrawalravi90 replied:
If a remote cluster's servers are part of PSI then num_svrs will include them. If not, then qdel does pbs_connect multiple times anyways, once for each cluster, and calls deljoblist once per cluster, so it should not be a problem.

agrawalravi90 replied:
on second thought, you are right, in its current form it get_num_servers() won't return the correct value for such a case, so I'll change it, thanks

agrawalravi90 replied:
ok, my whole broadcast logic depends on get_num_servers() returning the correct number of servers for the cluster. If and when we start supporting multi-cluster multi-server, we'll have to change it to take a connection descriptor and return the number of servers in the cluster, or add a new function to do so. So, I'll leave the code as it is for now, I don't want to make it confusing by using num_svrs in one for loop and conns[i] != NULL in another.

nithinj replied:
You dont have to use it inconsistently. You could replace all usages of your num_svrs check with conns[i] != NULL . num_svrs check is just the local version of the conns[i] != NULL .

agrawalravi90 commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -1200,11 +1199,94 @@ get_svr_inst_fd(int vfd, char *svr_inst_id)
+ * @retval 1 for yes, 0 for no
+ */
+int
+is_fd_virtual(int fd)
```
I prefer the word 'virtual' over 'cluster', virtual makes it clear that this fd is actually not real and the code needs to retrieve the real fd from elsewhere.  The API developer needs to call get_conn_svr_instances() if it's a virtual in order to get the real fd, so they do need to care right?

agrawalravi90 replied:
On second thought, I don't like is_fd_virtual either. How about we just call it "single_server_operation()" ? It'll return true if fd is real, false if fd is cluster/virtual/multi-server ?

nithinj commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -1200,11 +1199,94 @@ get_svr_inst_fd(int vfd, char *svr_inst_id)
+multi_svr_op(int fd)
+{
+	svr_conn_t **conns = get_conn_svr_instances(fd);
+	int num_svrs = get_num_servers();
```
Check the length of the array instead. If you want to check its single instance cluster, you can do `fd == conns[0]->fd`

nithinj reviewed (approved):
LGTM!

agrawalravi90 commented:
@arungrover  I've removed unknown jobs from the error list and simplified the broadcast code, please let me know how it looks.
@nithinj please take a look at the last commit as well, thanks.

arungrover reviewed (approved):
Thanks for making the change Ravi.
LGTM

agrawalravi90 commented:
@nithinj I've reverted the unknown job commit, let me know if it looks ok, thanks again for catching it.
@arungrover  i had to revert the last commit, we do need to return unknown jobs so that qdel can try to delete them on remote clusters, please let me know if it looks ok.

agrawalravi90 commented:
@nithinj  and @arungrover while running regressions i found an issue with the return code of qdel for error caxses, I fixed it in the latest commit, can you please take a look and merge this if all looks ok?

arungrover reviewed (approved):
LGTM
Sorry that we made you do that unnecessary change.
Once you rebase, i'll merge it.

bremanandjk commented:
@agrawalravi90, Please try to do the performance test after your final code change.

agrawalravi90 commented:
Latest performance results:
[ragrawal@blrentperf04 ~]$ qstat | wc -l
50002
[ragrawal@blrentperf04 ~]$ time qdel `qselect`

real	0m12.530s
user	0m0.145s
sys	0m0.043s

Regression:
![Screen Shot 2021-01-07 at 6 46 33 PM](https://user-images.githubusercontent.com/4574875/103957544-eea83c80-5118-11eb-9017-85fe3a3bfd5c.png)

agrawalravi90 commented:
@agrawalravi90  and @nithinj  it turned out that qdel used to actually report unknown jobs to the user. We broke this behavior with deljoblist. I fixed it in my latest commit. There was also an issue with deljoblist not colating the error lists properly. Please take a look at the latest commits.

nithinj reviewed (approved):
The recent changes look good to me!

arungrover reviewed (approved):
LGTM
